### PR TITLE
ZAPI-666: refactor handling of curl HEAD requests

### PR DESCRIPTION
**Pull request by misterdjules**

### CODE REVIEW

File: lib/vmapi.js [hunk 1]
```diff
@@ -170,15 +172,36 @@ VMAPI.prototype._initApis = function () {
             '*/*': formatJSON }
     });
 
+     // This allows VMAPI to respond to HEAD requests sent by curl with proper
+     // headers and prevents/fixes https://smartos.org/bugview/ZAPI-220.
+     this.server.pre(curlUserAgent());
```
Comment (joshwilsdon) [inline lib/vmapi.js]: It looks like this block got indented by one too many spaces.

File: lib/restify-plugins/curl-user-agent.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+// it to whatever restify branch VMAPI uses at that time and remove this local
+// copy.
+
+var assert = require('assert-plus');
+
+/**
+ * This basically exists for curl.  curl on HEAD requests usually
+ * just sits there and hangs, unless you explicitly set
+ * Connection:close.  And in general, you probably want to set
+ * Connection: close to curl anyway.
```
Comment (joshwilsdon) [inline lib/restify-plugins/curl-user-agent.js]: Would it make sense to add a test that explicitly ensures things still work fine when the curl client *has* passed 'Connection: close'? Maybe that's going too far? Is there any other possible value for 'Connection: ' that a client might ever set that we're silently overriding? Does that matter?

File: lib/vmapi.js (hunk 1 above)
Comment (misterdjules) [inline lib/vmapi.js]: @joshwilsdon Good catch :) Now I'm wondering why `make check` didn't catch it.

File: lib/restify-plugins/curl-user-agent.js (hunk 2 above)
Comment (misterdjules) [inline lib/restify-plugins/curl-user-agent.js]: > Would it make sense to add a test that explicitly ensures things still work fine when the curl client has passed 'Connection: close'

The `connection` header sent by VMAPI will be overriden to `close` only if the client has a user agent that matches `curl`. If a client uses `curl` and set the `connection` header to `close`, then the server will send a response with a `close` header, which seems fine.

However I agree it wouldn't hurt to add more test cases that send different `connection` header values when using `curl` and non-curl clients.

> Is there any other possible value for 'Connection: ' that a client might ever set that we're silently overriding? Does that matter?

Clients can send other values for the `connection` header, like `upgrade`. VMAPI doesn't support that though for now as far as I know, so I would not add a test case for this.

Comment (misterdjules) [inline lib/restify-plugins/curl-user-agent.js]: @joshwilsdon 

> However I agree it wouldn't hurt to add more test cases that send different connection header values when using curl and non-curl clients.

Added additional tests in latest commit.

**Comment by joshwilsdon** [comment]
lgtm

**Comment by joshwilsdon** [comment]
still looks good to me, including the latest commit.

**Comment by misterdjules** [comment]
Thank you, merged in 9491fc35527d214d1f1c781606b1b3ef99695bd3.

**Pull request closed without merging**